server/mongo_data_access.go

```diff
@@ -13,12 +13,60 @@ import (
 )
 
 // NewMongoDataAccessLayer returns an implementation of DataAccessLayer that is backed by a Mongo database
-func NewMongoDataAccessLayer(db *mgo.Database) DataAccessLayer {
-	return &mongoDataAccessLayer{Database: db}
+func NewMongoDataAccessLayer(db *mgo.Database, interceptors map[string]InterceptorList) DataAccessLayer {
+	return &mongoDataAccessLayer{
+		Database:     db,
+		Interceptors: interceptors,
+	}
 }
 
 type mongoDataAccessLayer struct {
-	Database *mgo.Database
+	Database     *mgo.Database
+	Interceptors map[string]InterceptorList
+}
+
+// Interceptor executes a function on a specified resource type immediately AFTER
+// the resource is modified in the database. To register an interceptor for ALL resource
+// types use a "*" as the resourceType.
+type Interceptor struct {
+	ResourceType string
+	Handler      InterceptorHandler
+}
+
+// InterceptorHandler is a function that is executed on a single FHIR resource
+type InterceptorHandler func(interface{})
+
+// InterceptorList is a list of interceptors registered for a given HTTP verb
+type InterceptorList []Interceptor
+
+// invokeInterceptors invokes the interceptor list for a particular HTTP verb and resource type.
+// Supported verbs are: POST, PUT, DELETE
+func (dal *mongoDataAccessLayer) invokeInterceptors(httpVerb, resourceType string, resource interface{}) {
+
+	var interceptors InterceptorList
+	if httpVerb == "POST" || httpVerb == "PUT" || httpVerb == "DELETE" {
+		interceptors = dal.Interceptors[httpVerb]
+		for _, interceptor := range interceptors {
+			if interceptor.ResourceType == resourceType || interceptor.ResourceType == "*" {
+				interceptor.Handler(resource)
+			}
+		}
+	}
+}
+
+// hasInterceptorsForVerbAndType checks if any interceptors are registered for a particular HTTP verb AND resource type
+func (dal *mongoDataAccessLayer) hasInterceptorsForVerbAndType(httpVerb, resourceType string) bool {
+
+	if len(dal.Interceptors[httpVerb]) > 0 {
+		interceptors := dal.Interceptors[httpVerb]
+		for _, interceptor := range interceptors {
```

Comment:
Totally _minor_ thing, but I think you can drop the containing `if` statement and assignment to `interceptors`.  Instead just loop over `range dal.Interceptors[httpVerb]`.  It doesn't buy you too much, but I think it is technically more efficient.

I wasn't going to even say anything, but if you have to come in here to fix code anyway, might as well do this too.